feat(query): make sanitizeProjection prevent projecting in paths deselected in the schema

vkarpov15 opened the pull request:
<!-- Thanks for submitting a pull request! Please provide enough information so that others can review your pull request. The two fields below are mandatory.

If you're making a change to documentation, do **not** modify a `.html` file directly. Instead, find the corresponding `.pug` file or test case in the `test/docs` directory. -->

**Summary**

[sanitizeProjection option](https://thecodebarbarian.com/whats-new-in-mongoose-5-13-sanitizeprojection.html) currently exists to prevent cases like `select({ name: '$password' })`, which would cause the `name` property to contain the value of the `password` property in newer versions of MongoDB.

While that is helpful, `sanitizeProjection` can do a bit more to prevent inclusion of sensitive data when the projection is potentially untrusted. With this PR, if `sanitizeProjection` is enabled, there is no way to project in a field that's deselected with `select: false` in the schema definition. If `password` has `{ type: String, select: false }` and `sanitizeProjection` is set, then `select('+password')`, `select('password')`, etc. will be ignored.

@hasezoey what do you think about this PR, is this a reasonable feature and do you think it's reasonable to ship in 8.5?

<!-- Explain the **motivation** for making this change. What problem does the pull request solve? -->

**Examples**

<!-- If this code fixes a bug or adds a new feature, provide an example demonstrating the change, unless you added a test. -->

hasezoey reviewed (approved):
> @hasezoey what do you think about this PR, is this a reasonable feature and do you think it's reasonable to ship in 8.5?

i think it is reasonable to ship it with 8.5.0.

---

i personally have not used this yet, but wouldnt this (according to the test cases) completely disable `find().select("+field -field")`, or is this just for fields which have a schema `select: false`?

as a side note, i had tried to search for option `sanitizeProjection` in the documentation, and the only mention of this is in `Query.prototype.setOptions` as `The following options are only for find(), findOne(), findById(), findOneAndUpdate(), findOneAndReplace(), findOneAndDelete(), and findByIdAndUpdate():`, no documentation about what this option does.

vkarpov15 commented:
@hasezoey re: `select('+field')`, just for `+` and just for fields that have `select: false` in the schema. Other fields are fine. I also added a `Query.prototype.sanitizeProjection` method in 0904a18 and added docs with examples to address the docs issue.
